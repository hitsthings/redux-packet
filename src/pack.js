import { bindActionCreators } from "redux";

/*
//usage of pack
packAll({
    fromContext: {
        selector: (state, contextA, contextB) => `itemsInContext`, // or a function returning this to make one per component
        actions: (contextA, contextB) => `actionsForContext`,
    }
})
const users = packAll({
    fromGroup: {
        selector: () => createStructuredSelector({
            users: usersForGroupSelector,
            usersLoading: usersLoadingForGroupSelector,
            usersAllLoaded: usersAllLoadedForGroupSelector,
            usersLoadingErrors: usersLoadingErrorsForGroupSelector,
        }),
        actions: (dispatch, group) => ({
            loadUsers: () => dispatch(loadUsersForGroup(group)),
            addUserToGroup: user => dispatch(addUserToGroup(user, group)),
        }), // OR a raw object like actions: { actionA: actionCreatorA, actionB: actionCreatorB }
    },
    byPermission: { ... }
})
*/

const memoizeFirstMapStateToProps = (selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext) => {
    if (executeContextSelectors) {
        let inner = (state, props) => {
            inner = (state, props) => selector(state, ...executeContextSelectors(props));
            if (props !== firstProps) { // nothing salvageable
                return inner(state, props); 
            }
            if (state !== firstState) { 
                return selector(state, ...firstContext); // same props, can memoize contexts
            }
            return firstValue;// same props and state, can return first value
        };
        return (state, props) => inner(state, props);
    }
    let inner = state => {
        inner = selector;
        return state === firstState ?
            firstValue :
            inner(state);
    };
    return state => inner(state);
};

const checkPacketDescriptor = (packetDescriptor, name, calledFnName, argExpectation) => {
    if (typeof packetDescriptor !== 'object' || packetDescriptor === null) {
        throw new TypeError(`${calledFnName} expects ${argExpectation}an object, but ${name} was not.`);
    }
    if (packetDescriptor.selector && typeof packetDescriptor.selector !== 'function') {
        throw new TypeError(`${calledFnName} expects selector to be a function, but ${name}.selector was not.`);
    }
    if (packetDescriptor.actions && (typeof packetDescriptor.actions !== 'function' && typeof packetDescriptor.actions !== 'object')) {
        throw new TypeError(`${calledFnName} expects actions to be a function or object, but ${name}.actions was not.`);
    }
};

const checkPacketMapAndGetNames = packetMap => {
    if (typeof packetMap !== 'object' || packetMap === null) {
        throw new TypeError('packAll(packetMap) expects packetMap to be an object.');
    }
    if (Array.isArray(packetMap)) {
        throw new TypeError('packAll(packetMap) expects packetMap to be an object, not an Array.');
    }
    const packetNames = Object.keys(packetMap);
    if (packetNames.length === 0) {
        throw new TypeError('packAll(packetMap) expected to receive an object with properties, but found no properties.')
    }
    return packetNames;
};

const selectorExecutor = selectors => {
    selectors = selectors.map(selector => typeof selector === 'function' ? selector : () => selector);
    return props => selectors.map(selector => selector(props));
};

const internalPack = (packetDescriptor, name) => {
    const PacketMaker = (...contextSelectors) => {
        const hasContext = contextSelectors.length !== 0;
        const executeContextSelectors = hasContext && selectorExecutor(contextSelectors);
        if ('minimumSelectorsExpected' in packetDescriptor && contextSelectors.length < packetDescriptor.minimumSelectorsExpected) {
            throw new Error(`The packet ${name} expects at least ${packetDescriptor.minimumSelectorsExpected} selectors to be provided.`);
        }
        const mapStateToProps = packetDescriptor.selector ? (firstState, firstProps) => {
            const firstContext = hasContext ? executeContextSelectors(firstProps) : [];
            let firstValue = packetDescriptor.selector(firstState, ...firstContext);
            if (typeof firstValue !== 'function') { // no factory, fully memoized first call
                return memoizeFirstMapStateToProps(packetDescriptor.selector, executeContextSelectors, firstState, firstProps, firstValue, firstContext);
            }
            if (hasContext) { // factory, but memoize first context if we can. Passing in a new object as firstState means we'll never return a memoized firstValue
                return memoizeFirstMapStateToProps(firstValue, executeContextSelectors, {}, firstProps, null, firstContext);
            }
            // factory, no context, no memoization
            return state => firstValue(state);
        } : null;
        const mapDispatchToProps = typeof packetDescriptor.actions === 'function' ?
            hasContext ?
                (dispatch, props) => packetDescriptor.actions(dispatch, ...executeContextSelectors(props)) :
                dispatch => packetDescriptor.actions(dispatch) :
                packetDescriptor.actions ?
                dispatch => bindActionCreators(packetDescriptor.actions, dispatch) :
                null;
        return {
            mapStateToProps,
            mapDispatchToProps,
        };
    };
    if (name) {
        Object.defineProperty(PacketMaker, 'name', {
            writable: false,
            enumerable: false,
            configurable: true,
            value: `PacketMaker(${name})`
        });
    }
    return PacketMaker;
};

export const pack = (packetDescriptor, name) => {
    checkPacketDescriptor(packetDescriptor, "packetDescriptor", "pack(packetDescriptor)", "");
    return internalPack(packetDescriptor, name);
};

export const packAll = packetMap => {
    const packetNames = checkPacketMapAndGetNames(packetMap);
    return packetNames.reduce((o,name) => {
        const packetDescriptor = packetMap[name];
        checkPacketDescriptor(packetDescriptor, `packetMap['${name}']`, "packAll(packetMap)", "each property to be ");
        o[name] = internalPack(packetDescriptor, name);
        return o;
    }, {});
};