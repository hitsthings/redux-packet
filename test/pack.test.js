import { packAll } from '../src/pack.js';

describe('packAll()', () => {
    describe('contract assertions', () => {
        describe('throws on invalid input', () => {
            [
                undefined,
                null,
                true,
                false,
                0,
                1,
                '',
                'string',
                ['array'],
                {},
                {a: true},
                {a: 1},
                {a: 'string'},
                {a: {selector:true}},
                {a: {selector:1}},
                {a: {selector:'string'}},
                {a: {actions:true}},
                {a: {actions:1}},
                {a: {actions:'string'}},
            ].forEach(value =>
                it(String(JSON.stringify(value)), () => {
                    expect(() => packAll(value).toThrowErrorMatchingSnapshot())
                })
            );
        });
        describe('accepts valid input', () => {
            [
                { a:{} },
                { a:{}, b: {} },
                { a:{ selector: () => {}} },
                { a:{ actions: {} } },
                { a:{ actions: () => {} } },
            ].forEach(value => {
                it(String(JSON.stringify(value)), () => {
                    expect(() => packAll(value)).not.toThrow();
                })
            })
        });
        
        it('returns an object of functions with keys matching the input', () => {
            expect(packAll({ a: {}})).toEqual({ a: expect.any(Function) });
        });
    });

    describe('minimumSelectorsExpected', () => {
        it('does not validate if not passed in', () => {
            const packed = packAll({ a: { }});
            expect(() => packed.a()).not.toThrow();
        })
        it('validates contextSelector count', () => {
            const packed = packAll({ a: { minimumSelectorsExpected: 1 }});
            expect(() => packed.a()).toThrowErrorMatchingSnapshot();
            expect(() => packed.a(() => {})).not.toThrow();
            expect(() => packed.a(() => {}, () => {})).not.toThrow();
        });
    });

    describe('packets', () => {
        describe('with context', () => {
            describe('simple', () => {
                let packedWithContext, contextProviderA, makeSelector, selector, loadActionCreator;
                const contextValue = 40;
                const mappedProps = { prop: 42 };
                const loadAction = { type: 'A' };
                beforeEach(() => {
                    contextProviderA = jest.fn().mockReturnValue(contextValue);
                    selector = jest.fn().mockReturnValue(mappedProps);
                    loadActionCreator = jest.fn().mockReturnValue(loadAction);
                    packedWithContext = packAll({
                        a: {
                            selector,
                            actions: {
                                load: loadActionCreator
                            },
                        }
                    }).a(contextProviderA);
                });
                it('returns all packet props', () => {
                    expect(packedWithContext).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = packedWithContext.mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state, contextValue);

                    const mapped = mapStateToProps(state, props);
                    expect(mapped).toBe(mappedProps);

                    expect(selector).toHaveBeenCalledTimes(1);
                    expect(contextProviderA).toHaveBeenCalledTimes(1);
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = packedWithContext.mapDispatchToProps(dispatch, props);
                    expect(mapped).toEqual({
                        load: expect.any(Function)
                    });
                    const loadParam = {};
                    const action = mapped.load(loadParam);
                    expect(loadActionCreator).lastCalledWith(loadParam);
                    expect(dispatch).lastCalledWith(loadAction);
                    expect(loadActionCreator).toHaveBeenCalledTimes(1);
                    expect(dispatch).toHaveBeenCalledTimes(1);
                    expect(contextProviderA).toHaveBeenCalledTimes(0);
                });
            });

            describe('complex', () => {
                let packed, contextProviderA, makeSelector, selector, actions;
                const contextValue = 40;
                const mappedProps = { prop: 42 };
                const mappedActions = {
                    load: () => ({ type: 'A' })
                };
                beforeEach(() => {
                    contextProviderA = jest.fn().mockReturnValue(contextValue);
                    selector = jest.fn().mockReturnValue(mappedProps);
                    makeSelector = jest.fn().mockReturnValue(selector);
                    actions = jest.fn().mockReturnValue(mappedActions);
                    packed = packAll({
                        a: {
                            selector: makeSelector,
                            actions,
                        }
                    });
                });
                it('returns all packet props', () => {
                    expect(packed.a(contextProviderA)).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls makeSelector and selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = packed.a(contextProviderA).mapStateToProps(state, props);
                    expect(makeSelector).lastCalledWith(state, contextValue);
                    expect(selector).not.toHaveBeenCalled();

                    const mapped = mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state, contextValue);
                    expect(mapped).toBe(mappedProps);

                    expect(makeSelector).toHaveBeenCalledTimes(1);
                    expect(selector).toHaveBeenCalledTimes(1);
                    expect(actions).not.toHaveBeenCalled();
                    expect(contextProviderA).toHaveBeenCalledTimes(1);
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = packed.a(contextProviderA).mapDispatchToProps(dispatch, props);
                    expect(actions).lastCalledWith(dispatch, contextValue);
                    expect(mapped).toBe(mappedActions);

                    expect(actions).toHaveBeenCalledTimes(1);
                    expect(selector).not.toHaveBeenCalled();
                    expect(makeSelector).not.toHaveBeenCalled();
                    expect(contextProviderA).toHaveBeenCalledTimes(1);
                });
            });
        });

        describe('without context', () => {
            describe('simple', () => {
                let packed, makeSelector, selector, loadActionCreator;
                const contextValue = 40;
                const mappedProps = { prop: 42 };
                const loadAction = { type: 'A' };
                beforeEach(() => {
                    selector = jest.fn().mockReturnValue(mappedProps);
                    loadActionCreator = jest.fn().mockReturnValue(loadAction);
                    packed = packAll({
                        a: {
                            selector,
                            actions: {
                                load: loadActionCreator
                            },
                        }
                    }).a();
                });
                it('returns all packet props', () => {
                    expect(packed).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = packed.mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state);

                    const mapped = mapStateToProps(state, props);
                    expect(mapped).toBe(mappedProps);

                    expect(selector).toHaveBeenCalledTimes(1);
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = packed.mapDispatchToProps(dispatch, props);
                    expect(mapped).toEqual({
                        load: expect.any(Function)
                    });
                    const loadParam = {};
                    const action = mapped.load(loadParam);
                    expect(loadActionCreator).lastCalledWith(loadParam);
                    expect(dispatch).lastCalledWith(loadAction);
                    expect(loadActionCreator).toHaveBeenCalledTimes(1);
                    expect(dispatch).toHaveBeenCalledTimes(1);
                });
            });

            describe('complex', () => {
                let packed, makeSelector, selector, actions;
                const mappedProps = { prop: 42 };
                const mappedActions = {
                    load: () => ({ type: 'A' })
                };
                beforeEach(() => {
                    selector = jest.fn().mockReturnValue(mappedProps);
                    makeSelector = jest.fn().mockReturnValue(selector);
                    actions = jest.fn().mockReturnValue(mappedActions);
                    packed = packAll({
                        a: {
                            selector: makeSelector,
                            actions,
                        }
                    });
                });
                it('returns all packet props', () => {
                    expect(packed.a()).toEqual({
                        mapStateToProps: expect.any(Function),
                        mapDispatchToProps: expect.any(Function),
                    });
                });
                it('calls makeSelector and selector in makeMapStateToProps', () => {
                    const state = {}, props = {};
                    const mapStateToProps = packed.a().mapStateToProps(state, props);
                    expect(makeSelector).lastCalledWith(state);
                    expect(selector).not.toHaveBeenCalled();

                    const mapped = mapStateToProps(state, props);
                    expect(selector).lastCalledWith(state);
                    expect(mapped).toBe(mappedProps);

                    expect(makeSelector).toHaveBeenCalledTimes(1);
                    expect(selector).toHaveBeenCalledTimes(1);
                    expect(actions).not.toHaveBeenCalled();
                });
                it('calls actions in mapDispatchToProps', () => {
                    const dispatch = jest.fn(), props = {};
                    const mapped = packed.a().mapDispatchToProps(dispatch, props);
                    expect(actions).lastCalledWith(dispatch);
                    expect(mapped).toBe(mappedActions);

                    expect(actions).toHaveBeenCalledTimes(1);
                    expect(selector).not.toHaveBeenCalled();
                    expect(makeSelector).not.toHaveBeenCalled();
                });
            });
        });
    });
});