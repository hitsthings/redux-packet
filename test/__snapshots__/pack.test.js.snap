// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packAll() minimumSelectorsExpected validates contextSelector count 1`] = `"The packet a expects at least 1 selector to be provided."`;
